packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2266,22 +2356,25 @@ export class KeyringController extends BaseController<
   async #removeEmptyKeyrings(): Promise<void> {
     this.#assertControllerMutexIsLocked();
     const validKeyrings: EthKeyring<Json>[] = [];
+    const validKeyringMetadata: KeyringMetadata[] = [];
 
     // Since getAccounts returns a Promise
     // We need to wait to hear back form each keyring
     // in order to decide which ones are now valid (accounts.length > 0)
 
     await Promise.all(
-      this.#keyrings.map(async (keyring: EthKeyring<Json>) => {
+      this.#keyrings.map(async (keyring: EthKeyring<Json>, index: number) => {
         const accounts = await keyring.getAccounts();
         if (accounts.length > 0) {
           validKeyrings.push(keyring);
+          validKeyringMetadata.push(this.state.keyringsMetadata[index]);
```

Comment:
We should use the internal `#keyringsMetadata` instead of the state here because there could be changes on the metadata array not yet committed to the state

```suggestion
          validKeyringMetadata.push(this.#keyringsMetadata[index]);
```